fvtest/compilerunittest/p/BinaryEncoder.cpp

```diff
@@ -0,0 +1,2856 @@
+        if (_words.size() != other._words.size())
+            throw std::invalid_argument("Attempt to XOR instructions of different sizes");
+
+        BinaryInstruction instr(_words);
+
+        for (size_t i = 0; i < _words.size(); i++)
+            instr._words[i] ^= other._words[i];
+
+        return instr;
+    }
+
+    bool operator==(const BinaryInstruction& other) const {
+        return other._words == _words;
+    }
+};
+
+std::ostream& operator<<(std::ostream& os, const BinaryInstruction& instr) {
+    os << "[";
+
+    for (size_t i = 0; i < instr.words().size(); i++) {
+        os << " 0x" << std::hex << std::setw(8) << std::setfill('0') << instr.words()[i];
+    }
+
+    os << " ]";
+
+    return os;
+}
+
+std::ostream& operator<<(std::ostream& os, const TR::InstOpCode::Mnemonic& opCode) {
+    os << TR::InstOpCode::metadata[opCode].name;
+    return os;
+}
+
+std::ostream& operator<<(std::ostream& os, const TR::RealRegister::RegNum& reg) {
+    if (reg >= TR::RealRegister::FirstGPR && reg <= TR::RealRegister::LastGPR)
+        os << "gr" << (static_cast<int>(reg) - static_cast<int>(TR::RealRegister::FirstGPR));
+    else if (reg >= TR::RealRegister::FirstFPR && reg <= TR::RealRegister::LastFPR)
+        os << "fp" << (static_cast<int>(reg) - static_cast<int>(TR::RealRegister::FirstFPR));
+    else if (reg >= TR::RealRegister::FirstVRF && reg <= TR::RealRegister::LastVRF)
+        os << "vr" << (static_cast<int>(reg) - static_cast<int>(TR::RealRegister::FirstVRF));
+    else if (reg >= TR::RealRegister::FirstCCR && reg <= TR::RealRegister::LastCCR)
+        os << "cr" << (static_cast<int>(reg) - static_cast<int>(TR::RealRegister::FirstCCR));
+    else
+        os << "???";
+
+    return os;
+}
+
+class PowerBinaryEncoderTest : public TRTest::CodeGenTest {
+public:
+    TR::Node* fakeNode;
+
+    uint32_t buf[64];
+
+    TR::RealRegister* createReg(TR_RegisterKinds kind, TR::RealRegister::RegNum reg, TR::RealRegister::RegMask mask) {
+        return new (cg()->trHeapMemory()) TR::RealRegister(kind, 0, TR::RealRegister::Free, reg, mask, cg());
+    }
+
+    PowerBinaryEncoderTest() {
+        fakeNode = TR::Node::create(TR::treetop);
+        cg()->setBinaryBufferStart(reinterpret_cast<uint8_t*>(&buf[0]));
+    }
+
+    BinaryInstruction getEncodedInstruction(size_t size, uint32_t off = 0) {
+        return std::vector<uint32_t>(&buf[off], &buf[off + (size / 4)]);
+    }
+
+    BinaryInstruction encodeInstruction(TR::Instruction *instr, uint32_t off = 0) {
+        instr->estimateBinaryLength(0);
+        cg()->setBinaryBufferCursor(reinterpret_cast<uint8_t*>(&buf[0]) + off);
+
+        instr->generateBinaryEncoding();
+        return getEncodedInstruction(instr->getBinaryLength(), off);
+    }
+
+    BinaryInstruction getBlankEncoding(TR::InstOpCode opCode) {
+        reinterpret_cast<uint32_t*>(opCode.copyBinaryToBuffer(reinterpret_cast<uint8_t*>(&buf[0])));
+
+        return getEncodedInstruction(4);
```

Comment:
I agree, but this magic number won't actually last too long. All instructions that currently make it to this point are 4 bytes long (and the only other real size is 0 bytes, which isn't that useful here). The refactor of the Power binary encoder will be adding a query to `TR::InstOpCode` to get the expected size for binary encoding.

Unfortunately, I obviously can't make use of this query here yet to clarify things since that would create a dependency issue, as the refactor is dependent on these tests being merged first.

Reply:
I see. I'm fine with this being fixed as part of the refactor.